src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -137,35 +139,96 @@ void initConfigAndTlog(String configFilePath) {
-                boolean externalConfigExists = Files.exists(externalConfig);
 
-                // if tlog is present, read the tlog first because the yaml config file may not be up to date
+                IOException tlogValidationError = null;
                 if (tlogExists) {
+                    try {
+                        ConfigurationReader.validateTlog(transactionLogPath);
+                    } catch (IOException e) {
+                        tlogValidationError = e;
+                    }
+                }
+
+                // if tlog is present, read the tlog first because the yaml config file may not be up to date
+                if (tlogExists && tlogValidationError == null) {
                     kernel.getConfig().read(transactionLogPath);
                 }
 
+                // tlog recovery logic if the main tlog isn't valid
+                if (tlogValidationError != null) {
+                    // Attempt to load from backup tlog file
+                    Path backupTlogPath = CommitableFile.getBackupFile(transactionLogPath);
+                    boolean backupValid = false;
+                    if (Files.exists(backupTlogPath)) {
+                        try {
+                            ConfigurationReader.validateTlog(backupTlogPath);
+                            backupValid = true;
+                        } catch (IOException e) {
+                            logger.atError().log("Backup transaction log at {} is invalid", backupTlogPath, e);
+                        }
+                    }
+
+                    if (backupValid) {
+                        logger.atError()
+                                .log("Transaction log {} is invalid and so is the backup at {}, will attempt to "
+                                                + "load configuration from {}", transactionLogPath, backupTlogPath,
+                                        bootstrapTlogPath, tlogValidationError);
+                        kernel.getConfig().read(backupTlogPath);
+                        readFromNonTlog = true;
+                    } else if (bootstrapTlogExists) {
+                        // If no backup or if the backup was invalid, then try loading from bootstrap
+                        logger.atError()
+                                .log("Transaction log {} is invalid and no usable backup exists, will attempt to load "
+                                                + "configuration from {}", transactionLogPath, bootstrapTlogPath,
+                                        tlogValidationError);
+                        kernel.getConfig().read(bootstrapTlogPath);
+                        readFromNonTlog = true;
+                    } else {
+                        // There are no files to load from
+                        logger.atError()
+                                .log("Transaction log {} is invalid and no usable backup exists", transactionLogPath,
+                                        tlogValidationError);
```

Comment:
Should the program exit here? Especially if `externalConfigExists` is false?

Reply:
Logging is basically always safe, but I'd be concerned about making us exit at this point. I think it could have some undesirable side-effects.